### DISCUSSION TITLE
Data Access Portal

### DISCUSSION THREAD

**Discussion by cimview** [ercot/api-specs#39] [General]
I am trying to use the Data Access Portal to pull a few months of the "LMPs by Settlement Point, Resource Node, and Trading Hubs".  Using the web interface, I can only download 500 files at a time because that is the pagination limit.  Since it's an every 5 minutes report, that's not quite two days worth.  And, after downloading just two pages worth, it always errors out with an HTTP 400 Bad Request even though all I did was to select a new page of results.  I also tried using the public API but it doesn't seem to return any data before 12/31/2023.

Does anyone have any tips on how to download lots of little files through the Data Access Portal interface, or maybe a way to automate it?

**Comment by mackermann-ercot** [comment]
Hello cimview, 

Our primary objective for the Public API project is to provide historical archives for our public EMIL products that are no longer available through our main ercot.com site. The file archive provides EMIL products for the last seven years (and in some cases much longer), however we only provide API access to the data within those products from mid December of 2023 year (during the Beta release).

However, you can download the NP6-788-CD archives in bulk by using the POST endpoint  `../api/public-reports/archive/NP6-788-CD/download` along with a JSON payload (and usual Authorization and Ocp-Apim-Subscription-Key headers) containing the document Id's for the archives you wish to retrieve.

**For example:**

To download three of the archive files matching four document id's, use the following payload in your POST request do the download endpoint mentioned above.

```
{
    "docIds": [1016533754, 1016532570, 1016531608]
}
```
Specifically for the NP6-788-CD product, you can download 1,000 per request (we do have to limit requests for some larger reports to a smaller number).

In order to get a specific list of document Id's, you can use the parent endpoint `../api/public-reports/archive/NP6-788-CD` with the desired `postDatetimeFrom` / `postDatetimeTo` range you wish to find.

**Example:**

A call to `https://api.ercot.com/api/public-reports/archive/NP6-788-CD` results in something similar to the following:

```
{
    "_meta": {
        "totalRecords": 1446987,
        "pageSize": 1000,
        "totalPages": 144699,
        "currentPage": 1,
        "query": {
            "parameterCount": 0,
            "parameters": {},
            "sortedBy": "postDatetime: DESC"
        }
    },
    "product": {
        "emilId": "NP6-788-CD",
        "name": "LMPs by Resource Nodes, Load Zones and Trading Hubs",
        "reportTypeId": 12300
    },
    "archives": [
        {
            "docId": 1016533754,
            "friendlyName": "LMPSROSNODENP6788_20240702_145511_csv",
            "postDatetime": "2024-07-02T14:55:14.000",
            "_links": {
                "endpoint": {
                    "href": "https://api.ercot.com/api/public-reports/archive/NP6-788-CD?download=1016533754"
                }
            }
        },
        {
            "docId": 1016532570,
            "friendlyName": "LMPSROSNODENP6788_20240702_145011_csv",
            "postDatetime": "2024-07-02T14:50:13.000",
            "_links": {
                "endpoint": {
                    "href": "https://api.ercot.com/api/public-reports/archive/NP6-788-CD?download=1016532570"
                }
            }
        },
        {
            "docId": 1016531608,
            "friendlyName": "LMPSROSNODENP6788_20240702_144511_csv",
            "postDatetime": "2024-07-02T14:45:13.000",
            "_links": {
                "endpoint": {
                    "href": "https://api.ercot.com/api/public-reports/archive/NP6-788-CD?download=1016531608"
                }
            }
        },
        ...
        ... etc
        ...
    ],
    "_links": {
        "self": {
            "href": "https://api.ercot.com/api/public-reports/archive/np6-788-cd"
        },
        "parent": {
            "href": "https://api.ercot.com/api/public-reports"
        }
    }
}
```

You can then gather the `docId` values you wish to retrieve and place them into your POST request.

Regards,
mackermann, ERCOT

**Comment by cimview** [comment]
Thank you so much!  

I was successful in using this API to get the data I needed.  This help was very much appreciated.
